File: pkg/video/hls/muxer.go
package hls

import (
	"bytes"
	"context"
	"errors"
	"io"
	"net/http"
	"nvr/pkg/log"
	"nvr/pkg/video/gortsplib"
	"sync"
	"time"
)

// MuxerFileResponse is a response of the Muxer's File() func.
type MuxerFileResponse struct {
	Status int
	Header map[string]string
	Body   io.Reader
}

// Muxer is a HLS muxer.
type Muxer struct {
	playlist   *playlist
	segmenter  *segmenter
	logf       log.Func
	videoTrack *gortsplib.TrackH264
	audioTrack *gortsplib.TrackMPEG4Audio

	mutex        sync.Mutex
	videoLastSPS []byte
	videoLastPPS []byte
	initContent  []byte
}

// ErrTrackInvalid invalid H264 track: SPS or PPS not provided into the SDP.
var ErrTrackInvalid = errors.New("invalid H264 track: SPS or PPS not provided into the SDP")

// NewMuxer allocates a Muxer.
func NewMuxer(
	ctx context.Context,
	segmentCount int,
	segmentDuration time.Duration,
	partDuration time.Duration,
	segmentMaxSize uint64,
	logf log.Func,
	videoTrack *gortsplib.TrackH264,
	audioTrack *gortsplib.TrackMPEG4Audio,
) *Muxer {
	playlist := newPlaylist(ctx, segmentCount)
	go playlist.start()

	m := &Muxer{
		playlist:   playlist,
		logf:       logf,
		videoTrack: videoTrack,
	}

	m.segmenter = newSegmenter(
		time.Now().UnixNano(),
		segmentDuration,
		partDuration,
		segmentMaxSize,
		videoTrack,
		audioTrack,
		m.playlist.onSegmentFinalized,
		m.playlist.partFinalized,
	)
	return m
}

// OnSegmentFinalizedFunc is injected by core.
type OnSegmentFinalizedFunc func([]SegmentOrGap)

// WriteH264 writes H264 NALUs, grouped by timestamp.
func (m *Muxer) WriteH264(ntp time.Time, pts time.Duration, nalus [][]byte) error {
	return m.segmenter.writeH264(ntp, pts, nalus)
}

// WriteAAC writes AAC AUs, grouped by timestamp.
func (m *Muxer) WriteAAC(pts time.Duration, au []byte) error {
	return m.segmenter.writeAAC(pts, au)
}

// File returns a file reader.
func (m *Muxer) File(
	name string,
	msn string,
	part string,
	skip string,
) *MuxerFileResponse {
	if name == "index.m3u8" {
		return primaryPlaylist(m.videoTrack, m.audioTrack)
	}

	if name == "init.mp4" {
		m.mutex.Lock()
		defer m.mutex.Unlock()

		sps := m.videoTrack.SPS

		if m.initContent == nil ||
			(!bytes.Equal(m.videoLastSPS, sps) ||
				!bytes.Equal(m.videoLastPPS, m.videoTrack.PPS)) {
			initContent, err := generateInit(m.videoTrack, m.audioTrack)
			if err != nil {
				m.logf(log.LevelError, "generate init.mp4: %w", err)
				return &MuxerFileResponse{Status: http.StatusInternalServerError}
			}
			m.videoLastSPS = m.videoTrack.SPS
			m.videoLastPPS = m.videoTrack.PPS
			m.initContent = initContent
		}

		return &MuxerFileResponse{
			Status: http.StatusOK,
			Header: map[string]string{
				"Content-Type": "video/mp4",
			},
			Body: bytes.NewReader(m.initContent),
		}
	}

	return m.playlist.file(name, msn, part, skip)
}

// VideoTrack returns the stream video track.
func (m *Muxer) VideoTrack() *gortsplib.TrackH264 {
	return m.videoTrack
}

// AudioTrack returns the stream audio track.
func (m *Muxer) AudioTrack() *gortsplib.TrackMPEG4Audio {
	return m.audioTrack
}

// WaitForSegFinalized blocks until a new segment has been finalized.
func (m *Muxer) WaitForSegFinalized() {
	m.playlist.waitForSegFinalized()
}

// NextSegment returns the first segment with a ID greater than prevID.
// Will wait for new segments if the next segment isn't cached.
func (m *Muxer) NextSegment(prevID uint64) (*Segment, error) {
	return m.playlist.nextSegment(prevID)
}

// VideoTimescale the number of time units that pass per second.
const VideoTimescale = 90000

// Sample .
type Sample interface {
	private()
}

// VideoSample Timestamps are in UnixNano.
type VideoSample struct {
	PTS        int64
	DTS        int64
	AVCC       []byte
	IdrPresent bool

	Duration time.Duration
}

func (VideoSample) private() {}

// AudioSample Timestamps are in UnixNano.
type AudioSample struct {
	AU  []byte
	PTS int64

	NextPTS int64
}

// Duration sample duration.
func (s AudioSample) Duration() time.Duration {
	return time.Duration(s.NextPTS - s.PTS)
}

func (AudioSample) private() {}

File: src/hls/muxer.rs
use crate::{
    init::generate_init,
    playlist::{primary_playlist, Playlist},
    segmenter::{H264Writer, Segmenter},
    HlsQuery,
};
use async_trait::async_trait;
use bytes::Bytes;
use common::{time::DurationH264, DynLogger, SegmentFinalized, TrackParameters};
use http::{HeaderName, HeaderValue, StatusCode};
use std::{
    collections::HashMap,
    fmt::Formatter,
    io::Cursor,
    sync::Arc,
    time::{SystemTime, UNIX_EPOCH},
};
use tokio::{io::AsyncRead, sync::Mutex};
use tokio_util::sync::CancellationToken;

impl std::fmt::Debug for MuxerFileResponse {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} {:?}", self.status, self.headers)
    }
}

#[derive(Debug)]
#[allow(clippy::module_name_repetitions)]
pub struct HlsMuxer {
    token: CancellationToken,
    playlist: Arc<Playlist>,
    params: TrackParameters,

    //videoLastSPS: []byte,
    //videoLastPPS: []byte,
    init_content: Mutex<Bytes>,
}

impl HlsMuxer {
    pub fn new(
        parent_token: &CancellationToken,
        logger: &DynLogger,
        segment_count: usize,
        segment_duration: DurationH264,
        part_duration: DurationH264,
        segment_max_size: u64,
        params: TrackParameters,
    ) -> (Self, H264Writer) {
        let token = parent_token.child_token();
        let playlist = Arc::new(Playlist::new(token.clone(), logger.clone(), segment_count));

        let now = i64::try_from(
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .expect("time went backwards")
                .as_nanos(),
        )
        .expect("time should fit i64");

        let segmenter = Segmenter::new(
            now,
            segment_duration,
            part_duration,
            segment_max_size,
            playlist.clone(),
        );

        let muxer = Self {
            token,
            playlist,
            params,
            init_content: Mutex::new(Bytes::new()),
        };
        (muxer, H264Writer::new(segmenter))
    }

    pub fn cancel(&self) {
        self.token.cancel();
    }

    pub async fn file(&self, name: &str, query: &HlsQuery) -> MuxerFileResponse {
        if name == "index.m3u8" {
            return primary_playlist(&self.params.codec);
        }

        if name == "init.mp4" {
            //sps := m.videoTrack.SPS

            if self.init_content.lock().await.is_empty()
            /*|| (!bytes.Equal(m.videoLastSPS, sps) ||
            !bytes.Equal(m.videoLastPPS, m.videoTrack.PPS))*/
            {
                let result = generate_init(&self.params);
                match result {
                    Ok(init_content) => *self.init_content.lock().await = init_content,
                    Err(_) => {
                        //m.logf(log.LevelError, "generate init.mp4: %w", err)
                        return MuxerFileResponse {
                            status: StatusCode::INTERNAL_SERVER_ERROR,
                            headers: None,
                            body: None,
                        };
                    }
                }
                //m.videoLastSPS = m.videoTrack.SPS
                //m.videoLastPPS = m.videoTrack.PPS
            }

            let body = Box::new(Cursor::new(self.init_content.lock().await.clone()));
            #[allow(clippy::unwrap_used)]
            return MuxerFileResponse {
                status: StatusCode::OK,
                headers: Some(HashMap::from([(
                    HeaderName::from_bytes(b"Content-Type").unwrap(),
                    HeaderValue::from_str("video/mp4").unwrap(),
                )])),
                body: Some(body),
            };
        }

        self.playlist.file(name, query).await
    }

    // Blocks until a new segment has been finalized.
    /*async fn wait_for_seg_finalized(&self) {
        self.playlist.wait_for_seg_finalized().await
    }*/
}

#[async_trait]
impl common::HlsMuxer for HlsMuxer {
    fn params(&self) -> &TrackParameters {
        &self.params
    }

    // Returns the first segment with a ID greater than prevID.
    // Will wait for new segments if the next segment isn't cached.
    async fn next_segment(&self, prev_id: u64) -> Option<Arc<SegmentFinalized>> {
        self.playlist.next_segment(prev_id).await
    }
}

#[async_trait]
pub trait NextSegmentGetter {
    async fn next_segment(&self, prev_id: u64) -> Option<Arc<SegmentFinalized>>;
}

// Response of the Muxer's File() fn.
#[allow(clippy::module_name_repetitions)]
pub struct MuxerFileResponse {
    pub status: StatusCode,
    pub headers: Option<HashMap<HeaderName, HeaderValue>>,
    pub body: Option<Box<dyn AsyncRead + Send + Unpin>>,
}

impl MuxerFileResponse {
    #[cfg(test)]
    #[allow(clippy::unwrap_used)]
    pub async fn print(mut self) -> String {
        use tokio::io::AsyncReadExt;
        let body = if let Some(body) = &mut self.body {
            let mut buf = "\n".to_owned();
            body.read_to_string(&mut buf).await.unwrap();
            buf
        } else {
            String::new()
        };
        format!("{}\n{:?}{}", self.status, self.headers, body)
    }
}

pub const MUXER_FILE_RESPONSE_CANCELLED: MuxerFileResponse = MuxerFileResponse {
    status: StatusCode::NOT_FOUND,
    headers: None,
    body: None,
};

pub const MUXER_FILE_RESPONSE_ERROR: MuxerFileResponse = MuxerFileResponse {
    status: StatusCode::INTERNAL_SERVER_ERROR,
    headers: None,
    body: None,
};

pub const MUXER_FILE_RESPONSE_BAD_REQUEST: MuxerFileResponse = MuxerFileResponse {
    status: StatusCode::BAD_REQUEST,
    headers: None,
    body: None,
};

pub const MUXER_FILE_RESPONSE_NOT_FOUND: MuxerFileResponse = MuxerFileResponse {
    status: StatusCode::NOT_FOUND,
    headers: None,
    body: None,
};
